packages/sources/universal-v3/src/utils/buildResponse.ts

```diff
@@ -0,0 +1,183 @@
+import type { AxiosRequestConfig } from 'axios'
+import type { ResponseGenerics, AdapterResponse } from '@chainlink/external-adapter-framework/util'
+import type { AdapterConfig } from '@chainlink/external-adapter-framework/config'
+import { customSettings } from '..'
+
+export interface Response extends ResponseGenerics {
+  Data: {
+    result: hexstring | ''
+    error: hexstring | ''
+    errorString?: string
+    details?: string
+    userHttpQueries?: AxiosRequestConfig[]
+  }
+  Result: 'success' | 'error'
+}
+
+export type SandboxOutput = SandboxSuccess | SandboxError
+
+type hexstring = `0x${string}`
+
+interface SandboxSuccess {
+  success: hexstring
+  userHttpQueries: AxiosRequestConfig[]
+}
+
+interface SandboxError {
+  error: {
+    name: string
+    message: string
+    details?: string
+    userHttpQueries?: AxiosRequestConfig[]
+  }
+}
+
+export class ResponseBuilder {
+  public maxResponseBytes: number
+
+  constructor(config: AdapterConfig<typeof customSettings>) {
+    this.maxResponseBytes = config.MAX_RESPONSE_BYTES
+  }
+
+  public buildResponse = (sandboxOutput: unknown): AdapterResponse<Response> => {
+    if (this.isValidSandboxError(sandboxOutput)) {
+      return this.buildSandboxUserErrorResponse(sandboxOutput)
+    }
+
+    if (this.isValidSandboxSuccess(sandboxOutput)) {
+      return {
+        result: 'success',
+        data: {
+          result: sandboxOutput.success,
+          error: '',
+          userHttpQueries: sandboxOutput.userHttpQueries,
+        },
+        statusCode: 200,
+        timestamps: {
+          providerDataReceived: 0,
+          providerDataRequested: 0,
+          providerIndicatedTime: 0,
+        },
+      }
+    }
+
+    return this.buildGenericUserErrorResponse('Invalid output from provided source code')
+  }
+
+  public buildSandboxUserErrorResponse = (
+    sandboxOutput: SandboxError,
+  ): AdapterResponse<Response> => {
+    const errorString = `${sandboxOutput.error.name}: ${sandboxOutput.error.message}`
+
+    return {
+      result: 'error',
+      data: {
+        result: '',
+        error: this.stringToValidHexString(errorString),
+        errorString,
+        details: sandboxOutput.error.details,
+        userHttpQueries: sandboxOutput.error.userHttpQueries,
+      },
+      statusCode: 200,
+      timestamps: {
+        providerDataReceived: 0,
+        providerDataRequested: 0,
+        providerIndicatedTime: 0,
```

Comment:
Apparently `undefined` is not an accepted type, but I can use `NaN`.

Since this is out of scope and I need to deliver by EoD, I am going to ignore this for now to make the deadline.  However, I will make a ticket and add it to the sprint.

Reply:
Undefined is only valid for the `providerIndicatedTime` (since not all providers specify one), but for the others we should be able to get values for all of them. Maybe it's better to keep them as 0 instead of NaN if the work will be done later, but it should be carried out at some point to avoid us having outliers in our metrics